common/funding_tx.c

```diff
@@ -53,3 +56,271 @@ struct bitcoin_tx *funding_tx(const tal_t *ctx,
+		weight += input_weight;
+
+		assert(amount_sat_add(total, *total, inputs[i]->input_satoshis));
+	}
+
+	return weight;
+}
+
+static size_t calculate_output_weights(struct output_info **outputs)
+{
+	size_t i, output_weights = 0, scriptlen;
+
+	for (i = 0; i < tal_count(outputs); i++) {
+		scriptlen = tal_bytelen(outputs[i]->script);
+		/* amount field + script + scriptlen varint */
+		output_weights += (8 + scriptlen + varint_size(scriptlen)) * 4;
+	}
+
+	return output_weights;
+}
+
+static size_t calculate_weight(struct input_info **opener_inputs,
+		               struct input_info **accepter_inputs,
+		               struct output_info **opener_outputs,
+		               struct output_info **accepter_outputs,
+			       struct amount_sat *opener_total,
+			       struct amount_sat *accepter_total)
+
+{
+	size_t weight;
+
+	/* version, input count, output count, locktime */
+	weight = (4 + 1 + 1 + 4) * 4;
+
+	/* add segwit fields: marker + flag */
+	weight += 1 + 1;
+
+	weight += calculate_input_weights(opener_inputs, opener_total);
+	weight += calculate_input_weights(accepter_inputs, accepter_total);
+
+	/* channel funding output: amount, len, scriptpubkey */
+	weight += (8 + 1 + BITCOIN_SCRIPTPUBKEY_P2WSH_LEN) * 4;
+
+	weight += calculate_output_weights(opener_outputs);
+	weight += calculate_output_weights(accepter_outputs);
+
+	return weight;
+}
+
+static const struct output_info *find_change_output(struct output_info **outputs)
+{
+	size_t i = 0;
+	for (i = 0; i < tal_count(outputs); i++) {
+		if (amount_sat_eq(outputs[i]->output_satoshis, AMOUNT_SAT(0)))
+			return outputs[i];
+	}
+	return NULL;
+}
+
+static struct amount_sat calculate_output_value(struct output_info **outputs)
+{
+	size_t i = 0;
+	struct amount_sat total = AMOUNT_SAT(0);
+
+	for (i = 0; i < tal_count(outputs); i++) {
+		assert(amount_sat_add(&total, total, outputs[i]->output_satoshis));
+	}
+	return total;
+}
+
+static void add_inputs(struct bitcoin_tx *tx, struct input_info **inputs)
+{
+	size_t i = 0;
+	for (i = 0; i < tal_count(inputs); i++) {
+		bitcoin_tx_add_input(tx, &inputs[i]->prevtx_txid, inputs[i]->prevtx_vout,
+				     BITCOIN_TX_DEFAULT_SEQUENCE,
+				     inputs[i]->input_satoshis, inputs[i]->script);
+	}
+}
+
+static void add_outputs(struct bitcoin_tx *tx, struct output_info **outputs,
+		        const struct amount_sat *change)
+{
+	size_t i = 0;
+	u8 *script;
+	struct amount_sat value;
+
+	for (i = 0; i < tal_count(outputs); i++) {
+		/* Is this the change output?? */
+		if (change && amount_sat_eq(outputs[i]->output_satoshis, AMOUNT_SAT(0))) {
+			/* If there's no change amount, we leave it out */
+			if (amount_sat_eq(*change, AMOUNT_SAT(0)))
+				continue;
+			value = *change;
+		} else
+			value = outputs[i]->output_satoshis;
+
+		script = tal_dup_arr(tx, u8, outputs[i]->script,
+				     tal_count(outputs[i]->script), 0);
+		bitcoin_tx_add_output(tx, script, value);
+	}
+}
+
+struct bitcoin_tx *dual_funding_funding_tx(const tal_t *ctx,
```

Comment:
I think this is missing a few BOLT quotes, too.  That way if the spec changes, you will get a hint as to where to change things.